File: 1638.count-substrings-that-differ-by-one-character.kt
// @leet start
class Solution {
    fun countSubstrings(s: String, t: String): Int {
      var f = Array(s.length + 1) { IntArray(t.length + 1) }
      for (i in s.indices) {
        for (j in t.indices) {
          if (s[i] == t[j]) {
            f[i + 1][j + 1] = f[i][j] + 1
          }
        }
      }
      var g = Array(s.length + 1) { IntArray(t.length + 1) }
      var ans = 0
      for (i in s.length - 1 downTo 0) {
        for (j in t.length - 1 downTo 0) {
          if (s[i] == t[j]) {
            g[i][j] = g[i + 1][j + 1] + 1
          }
          else {
            ans += (f[i][j] + 1) * (g[i + 1][j + 1] + 1)
          }
        }
      }
      return ans
        
    }
}
// @leet end

File: 1638.count-substrings-that-differ-by-one-character.js
// @leet start
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var countSubstrings = function(s, t) {
  var f = new Array(s.length + 1).fill().map(() => new Array(t.length + 1).fill(0));
  for (let i = 0; i < s.length; i++) {
    for (let j = 0; j < t.length; j++) {
      if (s[i] === t[j]) {
        f[i + 1][j + 1] = f[i][j] + 1;
      }
    }
  }
  var g = new Array(s.length + 1).fill().map(() => new Array(t.length + 1).fill(0));
  let res = 0;
  for (let i = s.length - 1; i >= 0; i--) {
    for (let j = t.length - 1; j >= 0; j--) {
      if (s[i] === t[j]) {
        g[i][j] = g[i + 1][j + 1] + 1;
      }
      else {
        res += (f[i][j] + 1) * g[i + 1][j + 1] + 1;
      }
    }
  }
  return res;
};
// @leet end

module.exports = countSubstrings
